### PR TITLE
Event with namespace support.

### PR DESCRIPTION
**Pull request by malash**
相比较`dataton`的区别：

- 更少的接口支持，只支持`get`与`set`，`state.load(xxx)`请使用`state.set(xxx)`替代
- 支持`cursor`链式调用和`state`、`cursor`同构。其实是上一个PR的内容：https://github.com/nofluxjs/state/pull/1
- 提供事件监听接口`listen`。不同的是，`listen`返回的是`Observable`，对其进行标准的`subscribe`和`unsubscribe`即可实现监听与取消。
- 支持`change`事件的部分监听，这也是重构`dataton`最重要的功能之一。因为调研后类似实现较少，目前依赖`EventEmitter2`的`namespace`实现。 相关讨论：https://github.com/nofluxjs/state/issues/2

未实现：

- 暂未实现`update`接口，原因是现在的实现对`set`的`value`是无感知的，即使有循环引用、函数等也不需要额外处理，但要实现`update`接口则必须进行深拷贝，要处理这些问题。
- 没有diff检查，无法区分修改是否真的造成了变动。原理同上。
- 因为缺少`load`、`mergeUpdate`等方法，批量修改会多次触发`change`事件，暂未解决。
- 待实现`push`、`pop`等无副作用方法

### CODE REVIEW

File: src/state.js [hunk 1]
```diff
@@ -38,5 +52,23 @@ export default class State {
       return this.cursor(subPath).set(value);
     }
     this.__store.write(this.__cursor, value);
+    this.__emitter.emit(['change', ...this.__cursor, '**'].join('.'), value);
+  }
+
+  currentCursor() {
+    return this.__cursor.join('.');
```
Comment (ssnau) [inline src/state.js]: why call `join`? 我觉得直接返回`this.__cursor`即可。数组是精确的表现形式，字符串有歧义。e.g. `['a.b', 'c'] ` doesn't equal to `a.b.c.`

File: test/state.js [hunk 2]
```diff
@@ -42,3 +45,33 @@ test('cursor set', t => {
...
+    .timeoutWith(OBSERVABLE_TIMEOUT, Observable.empty());
+});
+
+test('event emit with cursor', t => {
+  t.plan(3);
+  const state = new State();
+  process.nextTick(() => {
+    state.set('', 1);
+    state.set('a', 1);
+    state.set('a.b', 1);
```
Comment (ssnau) [inline test/state.js]: will `state.cursor('a').set('c', 2)` also raise event?

Comment (malash) [reply test/state.js]: yes and i fixed the test.
https://github.com/nofluxjs/state/pull/3/commits/50b840986794fd7a08f75090262837493e04fab3#diff-f13dfcc101d7d9daad638042913a3418R73

File: src/state.js (hunk 1 above)
Comment (malash) [reply src/state.js]: 这里疏忽了，其实这个接口是否需要暴露是存疑的（原本想复用这部分逻辑，后来不用了，也忘删了，所以也没有测试）。

这里存在两个问题：

一是`store`相当于内存，而`cursor`相当于指针，这是一个树状的结构结构。保留`__cursor`的作用是 ~~便于`set`操作时的指针处理（`JavaScript`里没有传址所以不如保存路径）~~ 实现`copy-on-write`更新祖先结点。但对于已经遍历到子树的指针，是否有必要对外暴露父节点信息呢？

二是`state`设计时**没有考虑**区分'a.b'与'a'.'b'的区别。虽然可以支持数组与字符串两种指针定位方式，但这很容易混淆，比如`state.cursor(['a', 'b.c', 'd'])`和`state.cursor('a').cursor('b.c').cursor('d')`是否一样呢？

P.S. 其实问题二的实现还有别的办法，比如使用转义（如`state.cursor('a..b')`）

Comment (ssnau) [reply src/state.js]: 其实保留set('a.b.c', 1)的写法主要是因为这样的代码可读性更高，对书写也更友好，并且99.999%的情况下不会有什么问题。如果是corner case，使用者自然会写成`['a.b', 'c']`。问题二建议不解决，但凡key中带`.`的，使用者写成数组即可。不用新学一套语法。

File: test/state.js [hunk 3]
```diff
@@ -42,3 +45,38 @@ test('cursor set', t => {
...
+});
+
+test('event emit with cursor', t => {
+  t.plan(4);
+  const state = new State();
+  process.nextTick(() => {
+    // should emit
+    state.set('', 1);
+    state.set('a', 1);
+    state.cursor('a').set('c', 1);
```
Comment (ssnau) [inline test/state.js]: should emit的是3个，但plan的是4?

Comment (malash) [reply test/state.js]: 注释错了，是4个

File: src/state.js (hunk 1 above)
Comment (malash) [reply src/state.js]: 可以，我需要修改`listen`方法，`eventemitter2`支持数组形式的`message`，正好符合要求。

以及我暂时去掉了`currentCursor`方法，如果有适用场景可以再加上（内部直接调用state.__cursor即可）。

Comment (malash) [reply src/state.js]: fixed https://github.com/nofluxjs/state/pull/3/commits/fcf86f9e095cc282536d43c9a301f04406fea1b2
